src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseApprovalService.cs

```diff
@@ -113,92 +113,155 @@ IEmailTemplateService emailTemplateService
+                    releaseToUpdate.ApprovalStatus = request.ApprovalStatus;
+                    releaseToUpdate.NextReleaseDate = request.NextReleaseDate;
+                    releaseToUpdate.NotifySubscribers = originalRelease.Version == 0 || (request.NotifySubscribers ?? false);
+                    releaseToUpdate.UpdatePublishedDate = request.UpdatePublishedDate ?? false;
+                    releaseToUpdate.PublishScheduled = request.PublishMethod == PublishMethod.Immediate
                         ? _dateTimeProvider.UtcNow
                         : request.PublishScheduledDate;
 
                     var releaseStatus = new ReleaseStatus
                     {
-                        Release = release,
+                        Release = releaseToUpdate,
                         InternalReleaseNote = request.InternalReleaseNote,
                         ApprovalStatus = request.ApprovalStatus,
                         CreatedById = _userService.GetUserId()
                     };
 
-                    return await ValidateReleaseWithChecklist(release)
-                        .OnSuccessDo(() => RemoveUnusedImages(release.Id))
-                        .OnSuccessVoid(async () =>
+                    return await
+                        ValidateReleaseWithChecklist(releaseToUpdate)
+                        .OnSuccess(() => RemoveUnusedImages(releaseToUpdate.Id))
+                        .OnSuccessDo(async () =>
                         {
-                            _context.Releases.Update(release);
+                            _context.Releases.Update(releaseToUpdate);
                             await _context.AddAsync(releaseStatus);
                             await _context.SaveChangesAsync();
-
-                            // Only need to inform Publisher if changing release approval status to or from Approved
-                            if (oldStatus == ReleaseApprovalStatus.Approved ||
-                                request.ApprovalStatus == ReleaseApprovalStatus.Approved)
-                            {
-                                await _publishingService.ReleaseChanged(
-                                    releaseId,
-                                    releaseStatus.Id,
-                                    request.PublishMethod == PublishMethod.Immediate
-                                );
-                            }
-
-                            switch (request.ApprovalStatus)
+                        })
+                        .OnSuccess(() =>
+                            SendEmailNotificationsAndInvites(request, releaseToUpdate)
+                            .OnSuccess(() => NotifyPublisher(releaseId, request, oldStatus, releaseStatus))
+                            // If a failure occurs in either the sending of the emails or the Publisher notification,
+                            // roll back changes to the Release and remove the added Release Status.
+                            .OnFailureDo(async _ =>
```

Comment:
Is there anything extra that could be mentioned here to explain why we aren't just delaying the saving until after both of those are successful? Without reading the PR description first, if someone hit this bit of code they might be tempted to try undoing this.

Is notifying the publisher likely to fail given that it's just a queue message drop? It looks like our code within that method would currently only fail gracefully if the release isn't found.

I think delaying the save until the notifications have been sent, then doing the save, and then notifying the publisher could have been another approach here, but happy to go with this!

Reply:
Thanks for pushing on this.

Turns out the root problem was that the ReleaseStatus, if being saved after the NotifyChange message was sent, would not have been assigned an Id yet and so the message to the Publisher contained an Guid.Empty ReleaseStatusId.

The fix is to assign the Id manually before sending the Publisher message.

As it turns out, the "rollback" tests can continue to assert that nothing's changed as they were before.

Reply:
Cool, it was great to spot that one!